### PR TITLE
Add inset support

### PR DESCRIPTION
**Pull request by puneet-pdx**
## Summary:
Adds support for ViewInsets to the composable `MapView()` function.

## Description:
- Adds a non-nullable `mapInsets` parameter to the `MapView()` function. The parameter is defaulted to an zero PaddingValues
- A micro app `MapInsetsApp` has been added to test & demo MapInsets on the MapView.

### CODE REVIEW

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt [hunk 1]
```diff
@@ -129,6 +135,17 @@ public fun MapView(
         }
     }
 
+    LaunchedEffect(mapInsets) {
+        // When this call is made in the AndroidView's update callback, ViewInsets are not applied
+        // on the mapview on initial load. So we set the ViewInsets here.
+        mapView.setViewInsets(
+            mapInsets.calculateStartPadding(layoutDirection).value.toDouble(),
+            mapInsets.calculateEndPadding(layoutDirection).value.toDouble(),
```
Comment (hud10837) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:143]: I think it's worth being explicit here, so just to confirm:
- `MapView.setViewInsets` accepts left and right parameters, not start/end.
- By using start/end here, we are essentially adding support for layout direction start/end and doing the swap ourselves. It looks like core does not do any work to handle RTL layouts, so it's okay for us to do this here. But maybe this should be added to doc

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt [hunk 2]
```diff
@@ -88,6 +92,7 @@ public fun MapView(
     viewLabelProperties: ViewLabelProperties = ViewLabelProperties(),
     selectionProperties: SelectionProperties = SelectionProperties(),
     onViewpointChanged: (() -> Unit)? = null,
+    mapInsets: PaddingValues = PaddingValues(),
```
Comment (hud10837) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:95]: Can you add `@param` doc for this?

File: microapps/MapInsetsApp/app/src/main/res/values/themes.xml
Comment (hud10837) [inline microapps/MapInsetsApp/app/src/main/res/values/themes.xml:1]: Is this file necessary

File: microapps/MapInsetsApp/app/src/main/res/values/colors.xml
Comment (hud10837) [inline microapps/MapInsetsApp/app/src/main/res/values/colors.xml:1]: Is this file necessary

File: microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                        label = { Text("Top") },
+                        keyboardOptions = KeyboardOptions(
+                            keyboardType = KeyboardType.Decimal,
+                            imeAction = ImeAction.Done
+                        ),
+                        keyboardActions = KeyboardActions(onDone = { updateInsets(); focusManager.clearFocus() })
+                    )
+                    OutlinedTextField(
+                        value = bottomText,
+                        modifier = Modifier.size(96.dp, 64.dp),
```
Comment (hud10837) [inline microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt:117]: Should this also be 90dp?
```suggestion
                        modifier = Modifier.size(90.dp, 64.dp),
```

**Review by hud10837 — CHANGES_REQUESTED** [review]
Looks good, a few comments

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt (hunk 1 above)
Comment (puneet-pdx) [reply toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:143]: We are not adding support for layout direction start/end and doing the swap ourselves. The way left/right and start/end works is the other way around for paddingValues. I tested it out Start/End returns the same values when the layout changes from Ltr to Rtl, while left and right values are swapped. 

```Kotlin
/**
 * The padding to be applied along the start edge inside a box: along the left edge if
 * the layout direction is LTR, or along the right edge for RTL.
 */
@Stable
fun PaddingValues.calculateStartPadding(layoutDirection: LayoutDirection) =
    if (layoutDirection == LayoutDirection.Ltr) {
        calculateLeftPadding(layoutDirection)
    } else {
        calculateRightPadding(layoutDirection)
    }

...

@Immutable
internal class PaddingValuesImpl(
    @Stable
    val start: Dp = 0.dp,
    @Stable
    val top: Dp = 0.dp,
    @Stable
    val end: Dp = 0.dp,
    @Stable
    val bottom: Dp = 0.dp
) : PaddingValues {
    override fun calculateLeftPadding(layoutDirection: LayoutDirection) =
        if (layoutDirection == LayoutDirection.Ltr) start else end

    override fun calculateTopPadding() = top

    override fun calculateRightPadding(layoutDirection: LayoutDirection) =
        if (layoutDirection == LayoutDirection.Ltr) end else start

    override fun calculateBottomPadding() = bottom
```
https://androidx.tech/artifacts/compose.foundation/foundation-layout/1.0.0-beta05-source/androidx/compose/foundation/layout/Padding.kt.html

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt (hunk 2 above)
Comment (puneet-pdx) [reply toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:95]: updated

File: microapps/MapInsetsApp/app/src/main/res/values/themes.xml
Comment (puneet-pdx) [reply microapps/MapInsetsApp/app/src/main/res/values/themes.xml:1]: These are generated from the template project and I think we are not removing them. They are included with all the micro app modules. If we want to remove them then we need to decide and remove them everywhere.

File: microapps/MapInsetsApp/app/src/main/res/values/colors.xml
Comment (puneet-pdx) [reply microapps/MapInsetsApp/app/src/main/res/values/colors.xml:1]: These are generated from the template project and I think we are not removing them. They are included with all the micro app modules. If we want to remove them then we need to decide and remove them everywhere.

File: microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt (hunk 3 above)
Comment (puneet-pdx) [reply microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt:117]: I reduced the size of others to 90.dp to make room for this one which has longer text that was getting wrapped. I test it again and it seems to work for 90.dp as well. So I have changed it to make them all the same.

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt (hunk 1 above)
Comment (hud10837) [reply toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:145]: I'm pretty convinced then that we should be passing left and right. From what you sent, left and right will abide by start and end

Comment (puneet-pdx) [reply toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt]: updated

**Review by hud10837 — APPROVED** [review]
👍

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt (hunk 1 above)
Comment (hud10837) [reply toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt]: 🙏

File: microapps/MapInsetsApp/.gitignore [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+/.idea/modules.xml
+/.idea/workspace.xml
+/.idea/navEditor.xml
+/.idea/assetWizardSettings.xml
+.DS_Store
+/build
+/captures
+.externalNativeBuild
+.cxx
+local.properties
```
Comment (gunt0001) [inline microapps/MapInsetsApp/.gitignore:15]: The app project needs a readme file with a brief description. See readme for LocationDisplayApp for an example:
https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/blob/feature-branches/geo-compose/microapps/LocationDisplayApp/README.md

File: microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+import androidx.compose.ui.text.input.ImeAction
+import androidx.compose.ui.text.input.KeyboardType
+import androidx.compose.ui.text.input.TextFieldValue
+import androidx.compose.ui.unit.dp
+import com.arcgismaps.mapping.ArcGISMap
+import com.arcgismaps.mapping.BasemapStyle
+import com.arcgismaps.toolkit.geocompose.MapView
+
+@Composable
+fun MainScreen() {
```
Comment (gunt0001) [inline microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt:51]: Needs some doc with a brief description. See LocationDisplayApp for an example:
https://github.com/Esri/arcgis-maps-sdk-kotlin-toolkit/blob/feature-branches/geo-compose/microapps/LocationDisplayApp/app/src/main/java/com/arcgismaps/toolkit/locationdisplayapp/screens/MainScreen.kt#L54-L61

**Review by gunt0001 — COMMENTED** [review]
@puneet-pdx - looks good, a few comments...

File: microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+            val updateInsets = {
+                mapInsets = PaddingValues(
+                    leftText.text.toDoubleOrNull()?.dp ?: 0.0.dp,
+                    topText.text.toDoubleOrNull()?.dp ?: 0.0.dp,
+                    rightText.text.toDoubleOrNull()?.dp ?: 0.0.dp,
+                    bottomText.text.toDoubleOrNull()?.dp ?: 0.0.dp
+                )
+            }
+
+            val focusManager = LocalFocusManager.current
```
[the remark this answers is not in the document]
Comment (puneet-pdx) [reply microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt:79]: yes, it a Composable invocation can only happen from the context of a composable function. The inline calls are not happening in composable functions.

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt [hunk 7]
```diff
@@ -103,6 +110,7 @@ public fun MapView(
     val lifecycleOwner = LocalLifecycleOwner.current
     val context = LocalContext.current
     val mapView = remember { MapView(context) }
+    val layoutDirection = LocalLayoutDirection.current
```
[the remark this answers is not in the document]
Comment (puneet-pdx) [reply toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:125]: No, it a Composable invocation can only happen from the context of a composable function. The inline calls are not happening in composable functions.

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt [hunk 8]
```diff
@@ -58,11 +62,13 @@ import kotlinx.coroutines.launch
...
  * @param locationDisplay the [LocationDisplay] used by the composable [com.arcgismaps.toolkit.geocompose.MapView]
+ * @param wrapAroundMode the [WrapAroundMode] to specify whether continuous panning across the international date line is enabled
  * @param geometryEditor the [GeometryEditor] used by the composable [com.arcgismaps.toolkit.geocompose.MapView] to create and edit geometries by user interaction.
  * @param mapViewInteractionOptions the [MapViewInteractionOptions] used by this composable [com.arcgismaps.toolkit.geocompose.MapView]
  * @param viewLabelProperties the [ViewLabelProperties] used by the composable [com.arcgismaps.toolkit.geocompose.MapView]
  * @param selectionProperties the [SelectionProperties] used by the composable [com.arcgismaps.toolkit.geocompose.MapView]
- * @param wrapAroundMode the [WrapAroundMode] to specify whether continuous panning across the international date line is enabled
+ * @param mapInsets the Inset values to control the active visible area, instructing the MapView to ignore parts that may be obstructed
+ * by overlaid UI elements and affecting the MapView's logical center, the reported visible area and the location display.
```
[the remark this answers is not in the document]
Comment (puneet-pdx) [reply toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:75]: updated

File: microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+                        value = bottomText,
+                        modifier = Modifier.size(90.dp, 64.dp),
+                        onValueChange = { text: TextFieldValue -> bottomText = text },
+                        label = { Text("Bottom") },
+                        keyboardOptions = KeyboardOptions(
+                            keyboardType = KeyboardType.Decimal,
+                            imeAction = ImeAction.Done
+                        ),
+                        keyboardActions = KeyboardActions(onDone = { updateInsets(); focusManager.clearFocus() })
+                    )
```
[the remark this answers is not in the document]
Comment (puneet-pdx) [reply microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt:131]: updated

File: microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+            var bottomText by remember { mutableStateOf(TextFieldValue("")) }
+
+            val updateInsets = {
+                mapInsets = PaddingValues(
+                    leftText.text.toDoubleOrNull()?.dp ?: 0.0.dp,
+                    topText.text.toDoubleOrNull()?.dp ?: 0.0.dp,
+                    rightText.text.toDoubleOrNull()?.dp ?: 0.0.dp,
+                    bottomText.text.toDoubleOrNull()?.dp ?: 0.0.dp
+                )
+            }
```
[the remark this answers is not in the document]
Comment (puneet-pdx) [reply microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt:77]: Do we really need to ? We are already remembering the values for left, right, top and bottom text and this is invoked when ever one of the value changes.

File: microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt (hunk 5 above)
Comment (puneet-pdx) [reply microapps/MapInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapinsetsapp/screens/MainScreen.kt:57]: added

File: microapps/MapInsetsApp/.gitignore (hunk 4 above)
Comment (puneet-pdx) [reply microapps/MapInsetsApp/.gitignore:15]: added

File: toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt [hunk 11]
```diff
@@ -66,6 +68,8 @@ import kotlinx.coroutines.launch
  * @param mapViewInteractionOptions the [MapViewInteractionOptions] used by this composable [com.arcgismaps.toolkit.geocompose.MapView]
  * @param viewLabelProperties the [ViewLabelProperties] used by the composable [com.arcgismaps.toolkit.geocompose.MapView]
  * @param selectionProperties the [SelectionProperties] used by the composable [com.arcgismaps.toolkit.geocompose.MapView]
+ * @param insets the inset values to control the active visible area, instructing the MapView to ignore parts that may be obstructed
+ * by overlaid UI elements and affecting the MapView's logical center, the reported visible area and the location display.
```
Comment (gunt0001) [inline toolkit/geo-compose/src/main/java/com/arcgismaps/toolkit/geocompose/MapView.kt:72]: remove full stop:
```suggestion
 * by overlaid UI elements and affecting the MapView's logical center, the reported visible area and the location display
```

File: microapps/MapviewInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapviewinsetsapp/screens/MainScreen.kt [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+        Button(onClick = {
+            insets = PaddingValues(0.0.dp, 0.0.dp, 0.0.dp, 0.0.dp)
+            leftText = TextFieldValue("")
+            rightText = TextFieldValue("")
+            topText = TextFieldValue("")
+            bottomText = TextFieldValue("")
+            focusManager.clearFocus()
+        }, Modifier.fillMaxWidth()) {
+            Text("Reset Insets")
+        }
```
Comment (gunt0001) [inline microapps/MapviewInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapviewinsetsapp/screens/MainScreen.kt:123]: Nitpick - the button could do with some outer padding.

**Comment by gunt0001** [comment]
@hud10837 - please take another look once comments are addressed.

File: microapps/MapviewInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapviewinsetsapp/screens/MainScreen.kt (hunk 12 above)
Comment (puneet-pdx) [reply microapps/MapviewInsetsApp/app/src/main/java/com/arcgismaps/toolkit/mapviewinsetsapp/screens/MainScreen.kt:123]: added padding

File: microapps/MapviewInsetsApp/README.md [hunk 13]
```diff
@@ -0,0 +1,12 @@
+# MapView Insets Micro-app
+
+This micro-app demonstrates the use of `Insets` with a composable `MapView`.
+
+![Screenshot](screenshot.png)
```
[the remark this answers is not in the document]
Comment (puneet-pdx) [reply microapps/MapviewInsetsApp/README.md:5]: updated

**Review by hud10837 — APPROVED** [review]

**Review by gunt0001 — APPROVED** [review]

**Pull request merged**